File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
N, M = map(int, input().split())
trees = list(map(int, input().split()))


left = 0
right = max(trees)
answer = 0

while left <= right:
    saw_blood = (left + right) // 2
    total = 0

    
    for height in trees:
        total += max(0, height - saw_blood)
    
    # 필요한만큼 나무를 취했다면 정답을 최신화
    if total >= M:
        answer = saw_blood
        left = saw_blood + 1
    else:
        right = saw_blood - 1

print(answer)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm> 

using namespace std;

int main() {
    int n;
    long long m; 
    cin >> n >> m;
    vector<int> v(n);

    for (int i = 0; i < n; i++) {
        cin >> v[i];
    }

    sort(v.begin(), v.end());

    int left = 0;
    int right = v[n-1];

    int answer = 0;
    while (left <= right) {
        int saw = (left + right) / 2;

        long long total = 0; 

        for (int tree : v) {
            total += max(tree - saw, 0);
        }

        if (total >= m) {
            answer = saw;
            left = saw + 1;
        } else {
            right = saw - 1;
        }
    }
    cout << answer << endl;
    return 0;
}
